### ISSUE TITLE
Weak load files

### ISSUE THREAD

**Issue by stipsan** [medialize/sass.js#24]
Any change we can make it so it's possible to pass a callback to Sass.writeFile, so that files only need to be loaded into memory whenever the actual file is loaded.
Working on a project that got a lot of deps to stuff like Bourbon, Foundation etc so it's hard to know what files to load and what not, so everything need to be standby.

**Comment by rodneyrehm** [comment]
More or less a duplicate of #13 

I've experimented with this in the past by hooking into the `stat` callback to see what libsass is trying to load. The problem here is that at the time of stat we immediately need the file, so we'd have to load it synchronously.

Alternatively we could - and this sounds equally bad - abort libsass execution upon stat and then restart when the previously requested file becomes available. This could repeat for a while since every newly available file could cause contain another import. This can be achieved with a fairly small amount of code.

Alternatively we could preflight-parse the source for `@import`s and download them all before running libsass compilation. This requires us to parse the source files and resolve relative paths. This requires slightly more code than the `stat` trick.

That said, we need a solution to this problem at some point.

**Comment by stipsan** [comment]
Thanks, I can help out on that if you can point me to the right direction for compiling a debug build of sass.js

**Comment by rodneyrehm** [comment]
1. clone the repo, checkout master
2. install emscripten
3. run `grunt build`
4. open console.html for unwrapped access to libsass.js

**Comment by stipsan** [comment]
Thanks, I got up and running last night. Compiled using  -s ALLOW_MEMORY_GROWTH=1 in order to take on sass projects in any size. It'll take some time for me to learn how emscripten works and how this can be done but at least I got the ball rolling :)

**Comment by rodneyrehm** [comment]
> Compiled using -s ALLOW_MEMORY_GROWTH=1 in order to take on sass projects in any size.

Why was that necessary?

---

When I looked at this stuff there was no real documentation for the Emscripten JS-API. I don't know if this is the case now. I learned how the internals work by investigating `console.html` using Chrome DevTools ability to dump the source of a function by simply running `FS.stat`. Here's how you can patch your way into the stat call, for example:

```js
// duckpunch a logger into the Emscripten JS-API stat call
FS.stat = (function(stat){
  return function(path) {
    console.log("stat", path);
    return stat.apply(this, arguments);
  };
})(FS.stat);

// write some files
Sass.writeFile('test.scss', '.world { background: red; }');
Sass.writeFile('foo/test.scss', '.world { background: green; }');

// see what libsass is trying to access
Sass.compile('@import "test.scss"; @import "foo/test.scss";');
```

procudes an output like:
```
stat /sass
stat /sass/foo
stat test.scss
stat _test.scss
…
```

maybe that helps…

**Comment by stipsan** [comment]
It was necessary as I'm testing sass.js in a project where I'm building a sass version of this: http://getuikit.com/docs/customizer.html
That is using Bourbon, Foundation and a in-house theme framework.
I don't have the privilige to share the code for that, but the challenge is that it's difficult to know what parts of the 3 libraries (Bourbon, Foundation and the custom theme framework) will be used in the currently loaded theme.
Thus the entirety of those three are loaded in case the theme imports them.
This is causing a lot of Sass.writeFile calls to happen (20+).

I don't know yet if it's simply the large amount of files or the complexity of the code in the files (color calculations, a lot of mixins, functions and math for some stuff that caused us such a slow-down we needed to switch to libsass on our stack), but I do know the error you get:
```
Uncaught Assertion: Cannot enlarge memory arrays. 
Adjust TOTAL_MEMORY (currently 52428800) or compile with ALLOW_MEMORY_GROWTH 
```
This error is not uncommon for emscripten projects: https://github.com/chandlerprall/Physijs/issues/64

It's very likely a memory leak caused by how writeFile works (loading it into memory?). But it's too early to say if this is a leak in sass.js or in libsass.
Either way, check out the emscripten docs: http://kripken.github.io/emscripten-site/docs/optimizing/Optimizing-Code.html
And the memory growth section. I think it's safe to assume sass.js can't know how much memory it will need in advance, so it may make prefect sense to always compile with this flag anyway?

Thanks for the FS.stat brief, it helps.

The emscripten filesystem api docs are much improved: http://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html#filesystem-api

FS.createLazyFile and FS.createPreloadedFile may be able to allow us to lazy load from XHR(?) as you got url parameters, not sure how we can prevent the files to load into memory but I'm investigating it at the moment.

**Comment by rodneyrehm** [comment]
> It's very likely a memory leak caused by how writeFile works (loading it into memory?). But it's too early to say if this is a leak in sass.js or in libsass.

This is the stuff I'd talk to @kripken about. If adding the flag *solved* the problem, a PR to get [memory growth](http://kripken.github.io/emscripten-site/docs/optimizing/Optimizing-Code.html#memory-growth) into the Makefile.patch would be nice.

> But it's too early to say if this is a leak in sass.js or in libsass.

considering that the error is thrown by emcsripten, if a leak exists, it is in libsass. sass.js itself is simply a convenience wrapper for the emscripten JS-API targeting libsass. It doesn't do anything signifcant itself.

a leak would mean memory is not cleaned up. Does your first compilation fail, or is it a subsequent attempt?

---

[FS.createLazyFile](http://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html#FS.createLazyFile) indeed looks interesting. But that still requires us to know all the SCSS files ahead of time, no?

**Comment by rodneyrehm** [comment]
I've been fiddling with [getting libsass 3.2 to build](https://github.com/medialize/sass.js/tree/libsass/3.2) and added the memory growth there.

**Comment by rodneyrehm** [comment]
I've just added `SassWorker._eval()` that allows running the following code locally using `worker.html`. This function is only there to help us play with `FS` and does not necessarily have to stay in the next release. `FS.createLazyFile()` only works in the WebWorker context because it requires synchronous XHR, which has been disabled in most browser's main threads.

```js
function registerFile() {
  this.FS.stat = (function(stat){
    return function(path) {
      console.log("stat", path);
      return stat.apply(this, arguments);
    };
  })(this.FS.stat);

  // Sass.writeFile() puts everything in the directory /sass/ (Sass._path)
  this._ensurePath('/sass/deeper/_some.scss');
  // resolution is relative to the worker's url (being in …/dist)
  this.FS.createLazyFile('/sass/deeper/', '_some.scss', '../scss/deeper/_some.scss', true, false);
}

function runCompile() {
  Sass.compile('@import "deeper/some";', console.log.bind(console))
}

Sass._eval(registerFile, runCompile);
```

This shows that [`FS.createLazyFile()`](http://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html#FS.createLazyFile) is indeed a viable option to register all *known* resources and sub-resources, without explicitly loading all of them up front:

```js
function registerFile() {
  // Sass.writeFile() puts everything in the directory /sass/ (Sass._path)
  this._ensurePath('/sass/example.scss');
  this._ensurePath('/sass/deeper/_some.scss');
  // resolution is relative to the worker's url (being in …/dist)
  this.FS.createLazyFile('/sass/', 'example.scss', '../scss/example.scss', true, false);
  this.FS.createLazyFile('/sass/', '_importable.scss', '../scss/_importable.scss', true, false);
  this.FS.createLazyFile('/sass/deeper/', '_some.scss', '../scss/deeper/_some.scss', true, false);
}

function runCompile() {
  Sass.compile('@import "example";', console.log.bind(console))
}

Sass._eval(registerFile, runCompile);
```

The next question is how to remove the requirement to do know all resources in advance.

**Comment by rodneyrehm** [comment]
ok, [the libsass 3.2 branch](https://github.com/medialize/sass.js/tree/libsass/3.2) has reached a working state. do you mind giving it a shot and see if it solves your problem? Note the added methods `Sass.lazyFiles()` and `Sass.preloadFiles()`.

Was there anything else in this issue other than allowing memory growth and registering larger batches of files? If not, I assume this issue can be closed once libsass 3.2 is released and the integration branch is merged into master.

**Comment by stipsan** [comment]
Hey, thanks for the comprehensive research.
I'll run tests and report my results.

I did encounter possible issues, one being failing to activate asm.js in firefox and some performance issues related to project size. But I'll create separate issues for those if they're valid and not just my setup.

You assume correctly.

**Comment by rodneyrehm** [comment]
I recently stumbled across the libsass [Importer API](https://github.com/sass/libsass/wiki/API-Sass-Importer) and gave [bridging it to JS world](https://github.com/medialize/sass.js/tree/libsass/3.2-importer#custom-file-loading) a shot. First results are in the branch [libsass/3.2-importer](https://github.com/medialize/sass.js/tree/libsass/3.2-importer). It *works*, but I guess the internals could be improved a bit.

**Comment by stipsan** [comment]
Forgot to get back to you, but better late than never, right? :)

`sass.preloadFiles` provided awesome performance after initial setup, but was a lot of work to implement support for updating the files on the client whenever they were changed on the server.
`sass.lazyFiles` was super easy to implement as we could test it outside the worker.
`sass.importer` is exactly what we needed. It allowed us to implement a separate mechanism for downloading the files from server, storing them in sessionStorage and updating them whenever they change on the server without blocking the sass compile. This gave us a really performant solution that supports "livereloading" :D 

Very nice work, and thanks again for implementing this! :+1:
